### PR TITLE
`MastForest` serialization

### PR DESCRIPTION
**Pull request by plafer**
Builds on #1359

Implements serialization of `MastForest`, as described in https://github.com/0xPolygonMiden/compiler/pull/132 (and taking into account @bobbinth's comments).

Notably, `MastNodeType` has been implemented to fit into 8 bytes instead of the proposed 9.

### CODE REVIEW

File: core/src/mast/serialization/info.rs [hunk 1]
```diff
@@ -0,0 +1,456 @@
...
+            MastNodeType::Loop { body_id: body } => Self::encode_u32_payload(*body),
+            MastNodeType::Block { len } => Self::encode_u32_payload(*len),
+            MastNodeType::Call { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::SysCall { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::Dyn => [0; 8],
+            MastNodeType::External => [0; 8],
+        }
+    }
+
+    fn encode_join_or_split(left_child_id: u32, right_child_id: u32) -> [u8; 8] {
```
Comment (plafer) [inline core/src/mast/serialization/info.rs:258]: The encoding of `JOIN` and `SPLIT` is a bit heavy; I can either document it in a comment here, or in the document that will describe the serialization format of `MastForest` (or both).

File: core/src/mast/serialization/info.rs [hunk 2]
```diff
@@ -0,0 +1,456 @@
...
+///
+/// The serialized representation of [`MastNodeInfo`] is guaranteed to be fixed width, so that the
+/// nodes stored in the `nodes` table of the serialzied [`MastForest`] can be accessed quickly by
+/// index.
+#[derive(Debug)]
+pub struct MastNodeInfo {
+    ty: MastNodeType,
+    offset: DataOffset,
+    digest: RpoDigest,
+}
```
Comment (bobbinth) [inline core/src/mast/serialization/info.rs:25]: Do we actually need to track the offset here? My understanding is that `offset` is useful only for basic blocks and for everything else it is set to 0. If so, we could probably move the offset into `MastNodeType` to make it look like so:
```rust
#[repr(u8)]
pub enum MastNodeType {
    Join {
        left_child_id: u32,
        right_child_id: u32,
    } = JOIN,
    Split {
        if_branch_id: u32,
        else_branch_id: u32,
    } = SPLIT,
    Loop {
        body_id: u32,
    } = LOOP,
    Block {
        /// Offset of the basic block in the data segment
        offset: u32,
        /// The number of operations and decorators in the basic block
        len: u32,
    } = BLOCK,
    Call {
        callee_id: u32,
    } = CALL,
    SysCall {
        callee_id: u32,
    } = SYSCALL,
    Dyn = DYN,
    External = EXTERNAL,
}
```

File: core/src/operations/mod.rs [hunk 3]
```diff
@@ -441,6 +443,148 @@ pub enum Operation {
     RCombBase,
 }
 
+pub enum OperationData {
+    Felt(Felt),
+    U32(u32),
+    None,
+}
```
Comment (bobbinth) [inline core/src/operations/mod.rs:450]: If we implement, `Serializable`/`Deserializable` for `Operation` as suggested in one of the previous comments, we probably won't need this struct.

File: core/src/operations/mod.rs [hunk 4]
```diff
@@ -441,6 +443,148 @@ pub enum Operation {
...
+impl Operation {
+    /// Builds an operation from its opcode and inline data (if any).
+    pub fn with_opcode_and_data(
+        opcode: u8,
+        data: OperationData,
+    ) -> Result<Self, DeserializationError> {
+        match opcode {
+            0b0000_0000 => Ok(Self::Noop),
+            0b0000_0001 => Ok(Self::Eqz),
+            0b0000_0010 => Ok(Self::Neg),
```
Comment (bobbinth) [inline core/src/operations/mod.rs:462]: This is rather error-prone. I wonder if we should just assign discriminant values in the `Operation` enum itself. This would allow us to get rid of this method and also `Operation::op_code()` would become obsolete.

If the above changes make sense, I probably would do them in a separate PR, merge that PR, and then rebase this PR from that one.

File: core/Cargo.toml [hunk 5]
```diff
@@ -28,9 +28,10 @@ std = [
 
 [dependencies]
 math = { package = "winter-math", version = "0.9", default-features = false }
-#miden-crypto = { version = "0.9", default-features = false }
 miden-crypto = { git = "https://github.com/0xPolygonMiden/crypto", branch = "next", default-features = false }
 miden-formatting = { version = "0.1", default-features = false }
+num-derive = "0.4.2"
+num-traits = "0.2.19"
```
Comment (bobbinth) [inline core/Cargo.toml:34]: nit: let's not specify patch versions.

File: core/src/mast/serialization/mod.rs [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+
+                mast_forest.add_node(node);
+            }
+
+            for root in roots {
+                mast_forest.make_root(root);
+            }
+
+            mast_forest
+        };
```
Comment (bobbinth) [inline core/src/mast/serialization/mod.rs:161]: Seems like we are iterating over node infos twice here: first time as we are deserializing them and second time when we try to deserialize the associated data. I think if we altered how we serialize things a bit, we could avoid this. Specifically:

Currently, we have the following order:
```
[magic, version, num_nodes, roots, node_infos, string_table, data]
```
But maybe the following is better:
```
[magic, version, num_nodes, roots, string_table, data, node_infos]
```
This way, we deserialize string table and data first and so we have everything in memory and would need to iterate over node infos only once.

This will make serialization a bit less efficient - but I think we care more about deserialization speed then serialization speed.

**Review by bobbinth — COMMENTED** [review]
Thank you! Looks good! Not a full review yet (though, I went through probably 80% of the code) - but I did leave some comments inline.

File: core/src/mast/serialization/basic_block_data_builder.rs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+        self.data.push(operation.op_code());
+
+        // For operations that have extra data, encode it in `data`.
+        match operation {
+            Operation::Assert(value) | Operation::MpVerify(value) => {
+                self.data.extend_from_slice(&value.to_le_bytes())
+            }
+            Operation::U32assert2(value) | Operation::Push(value) => {
+                self.data.extend_from_slice(&value.as_int().to_le_bytes())
+            }
```
[the remark this answers is not in the document]
Comment (Overcastan) [reply core/src/mast/serialization/basic_block_data_builder.rs:70]: Yes, we should change it to `u32`, there is no reason (known to me) to make imm a `Felt`.

I noticed this in [this old PR](https://github.com/0xPolygonMiden/miden-vm/pull/1274#issuecomment-1981825865), but after assembly refactoring was done we started to parse `u32assert2` with `u32` immediate value so I mistakenly thought that this issue was fixed, although I forget to check that internals of this operation didn't change and we still use `Felt` in it. 

I will create an issue for this. Would it be better to create a PR targeting this branch and not next?

[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/mast/serialization/basic_block_data_builder.rs:70]: Targeting this branch is probably better.

[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs]: Fixed in #1382 - thank you @Fumuran!

File: core/src/mast/serialization/basic_block_data_decoder.rs [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            data,
+            data_reader,
+            strings,
+        }
+    }
+}
+
+/// Mutators
+impl<'a> BasicBlockDataDecoder<'a> {
+    pub fn decode_operations_and_decorators(
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs:32]: >Question: why not implement Serializable and Deserializable for Operation instead of having separate functions here? It seems to me that encapsulating this logic there would be cleaner.

Oh yes much cleaner that way indeed. Fixed.

> Though, maybe we can still implement these traits for some of the internal decorator types (e.g., AdviceInjector, Debug, Event etc. - with AssemblyOp being the only exception).

This would be incompatible with my current strategy that "linearizes" the `Decorator` enums into one big enum `EncodedDecoratorVariant`. It would be a bigger change, and would still be messy, so I don't think it's worth it?

File: core/src/mast/serialization/basic_block_data_builder.rs [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+#[derive(Debug, Default)]
+struct StringTableBuilder {
+    table: Vec<StringRef>,
+    str_to_index: BTreeMap<RpoDigest, StringIndex>,
+    strings_data: Vec<u8>,
+}
+
+impl StringTableBuilder {
+    pub fn add_string(&mut self, string: &str) -> StringIndex {
+        if let Some(str_idx) = self.str_to_index.get(&Rpo256::hash(string.as_bytes())) {
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs:258]: Switched to using `Blake3_256`

File: core/src/mast/serialization/basic_block_data_builder.rs [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+                    .strings_data
+                    .len()
+                    .try_into()
+                    .expect("strings table larger than 2^32 bytes"),
+                len: string.len().try_into().expect("string larger than 2^32 bytes"),
+            };
+            let str_idx = self.table.len();
+
+            self.strings_data.extend(string.as_bytes());
+            self.table.push(str_ref);
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs:275]: The strings that we encode have a lot of repetition, (i.e. `AssemblyOp.context_name` is going to be the same for a bunch of operations, and similarly for `AssemblyOp.op`). If we stored the string length in the data buffer, we'd be duplicating that length byte (assuming a small string) over and over. Given that most strings are duplicates of one another, I suspect these length bytes could add up to a nontrivial amount. Storing the length byte in `StringRef` removes that duplication.

File: core/src/mast/serialization/decorator.rs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+///
+/// This is effectively equivalent to a set of constants, and designed to convert between variant
+/// discriminant and enum variant conveniently.
+#[derive(FromPrimitive, ToPrimitive)]
+#[repr(u8)]
+pub enum EncodedDecoratorVariant {
+    AdviceInjectorMerkleNodeMerge,
+    AdviceInjectorMerkleNodeToStack,
+    AdviceInjectorUpdateMerkleNode,
+    AdviceInjectorMapValueToStack,
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/decorator.rs:16]: Good idea. However, I moved the bit logic to the `EncodedDecoratorVariant::discriminant()` and `EncodedDecoratorVariant::from_discriminant()` methods, as writing out bit expressions explicitly is a bit error-prone.

File: core/src/operations/mod.rs (hunk 4 above)
Comment (plafer) [reply core/src/operations/mod.rs:463]: #1390

File: core/src/mast/serialization/basic_block_data_builder.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/mast/serialization/basic_block_data_builder.rs:159]: Maybe I'm misunderstanding how the strings table works, but don't we store only a single copy of a given string? If so, I was thinking that the length would be stored together with that string - so, there would be only one copy of the length stored as well.

File: core/src/mast/serialization/info.rs (hunk 2 above)
Comment (bobbinth) [reply core/src/mast/serialization/info.rs]: I think we might still needed. My understanding is that the current code assumes that we are deserializing the entire MAST forest - and in this case, we always know the location of the next basic block data implicitly.

However, I think one of the goals of this format is to make it possible to deserialize parts of the forest (no need to implement this right now, but we may want to have support for this in the format itself). For example, we may start with one of the roots and then "jump around" the file to read only the nodes relevant to that root. We can already do this for all nodes because `NodeType` is fixed size (so, given a node index, we can compute the location of this node without having to read the entire `nodes` data), but without associating offset into the data section for basic blocks, I'm not sure how we'd do it.

Comment (bobbinth) [reply core/src/mast/serialization/info.rs]: > This is a happy accident that saves us 4 bytes per `MastNode` 😄.

If we encode the offset into the `MastNodeType` - we would also save on these 4 bytes, right?

File: core/src/operations/mod.rs (hunk 3 above)
Comment (plafer) [reply core/src/operations/mod.rs]: merged `next`, and removed `OperationData`

File: core/src/mast/serialization/basic_block_data_builder.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs:159]: Oh no, I'm re-reading what I wrote and I think I was just confused... Applied your suggestion.

I kept the `StringRef` type, but it could be potentially cleaner to just use `DataOffset` (which makes it clear that a string reference is simply an offset into the data buffer)? I'm happy with both approaches.

File: core/src/mast/serialization/info.rs (hunk 2 above)
Comment (plafer) [reply core/src/mast/serialization/info.rs]: > However, I think one of the goals of this format is to make it possible to deserialize parts of the forest (no need to implement this right now, but we may want to have support for this in the format itself).

Ah yes you're totally right, I forgot about that.

> If we encode the offset into the MastNodeType - we would also save on these 4 bytes, right?

Absolutely; at the time of writing I somehow forgot that your initial suggestion had the exact same benefit, sorry about that.

I implemented the suggestion. The only nit I'm not happy about with the current implementation is that we currently ignore the offset stored in the `Block`, and use the one implicit in the `SliceReader`. However, someone could technically move blocks around during serialization, such that they're not in the order we expect. So at the very least, we should confirm that the block's offset corresponds to our `SliceReader`'s offset, which would require adding a method `SliceReader::position(&self)` to `SliceReader`.

File: core/src/mast/serialization/basic_block_data_decoder.rs [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        &mut self,
+        num_to_decode: u32,
+    ) -> Result<(Vec<Operation>, DecoratorList), DeserializationError> {
```
Comment (bobbinth) [inline core/src/mast/serialization/basic_block_data_decoder.rs:34]: One approach to address the issue you mentioned in https://github.com/0xPolygonMiden/miden-vm/pull/1370#discussion_r1674149012 is to instantiate a separate instance of `SliceReader` for each invocation of this method. The method signature would then take an offset as well and it could look like so:
```rust
pub fn decode_operations_and_decorators(
    &self,
    offset: DataOffset,
    num_to_decode: u32,
) -> Result<(Vec<Operation>, DecoratorList), DeserializationError> {
        let mut operations: Vec<Operation> = Vec::new();
        let mut decorators: DecoratorList = Vec::new();

        let mut data_reader = SliceReader::new(&data[offset as usize..]);
        for _ in 0..num_to_decode {
            let first_byte = data_reader.peek_u8()?;
```
Then, we also can get rid of the `data_reader` field from the `BasicBlockDecoder` struct. It could look something like this:
```rust
pub struct BasicBlockDataDecoder<'a> {
    data: &'a [u8],
    strings: &'a [StringRef],
}
```

**Review by bobbinth — CHANGES_REQUESTED** [review]
Looks good! Thank you! I left more comments inline. Most are pretty simple and the one about avoiding recomputing node hashes (which would probably require nontrivial amount of work) can be addressed in a follow-up PR.

One other thing: after this PR we are implicitly assuming that `NoteIndex` is a 30-bit value. Should we enforce this explicitly? (i.e., trying to construct a `NodeIndex` with a value greater than $2^{30} - 1$ should result in an error).

File: core/src/mast/serialization/info.rs (hunk 2 above)
Comment (bobbinth) [reply core/src/mast/serialization/info.rs]: > The only nit I'm not happy about with the current implementation is that we currently ignore the offset stored in the `Block`, and use the one implicit in the `SliceReader`.

I described a potential solution for this in https://github.com/0xPolygonMiden/miden-vm/pull/1370#discussion_r1674755804.

File: core/src/mast/serialization/basic_block_data_builder.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/mast/serialization/basic_block_data_builder.rs]: Yeah - I think using `DataOffset` is probably a bit cleaner.

File: core/src/operations/mod.rs [hunk 13]
```diff
@@ -916,15 +916,10 @@ impl Deserializable for Operation {
...
             OPCODE_U32ASSERT2 => {
-                let err_code_le_bytes: [u8; 8] = source.read_array()?;
-                let err_code_u64 = u64::from_le_bytes(err_code_le_bytes);
-                let err_code_felt = Felt::try_from(err_code_u64).map_err(|_| {
-                    DeserializationError::InvalidValue(format!(
-                        "Operation associated data doesn't fit in a field element: {err_code_u64}"
-                    ))
-                })?;
+                let err_code_le_bytes: [u8; 4] = source.read_array()?;
+                let err_code = u32::from_le_bytes(err_code_le_bytes);
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/operations/mod.rs:920]: My main concern was that `ByteReader` wouldn't enforce little-endian format; but the docstring for `read_u32()` does enforce that property, so yes let's use `source.read_u32()`.

Fixed.

File: core/src/mast/serialization/info.rs [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+use super::{basic_block_data_decoder::BasicBlockDataDecoder, DataOffset};
+
+// MAST NODE INFO
+// ================================================================================================
+
+/// Represents a serialized [`MastNode`], with some data inlined in its [`MastNodeType`].
+///
+/// In the case of [`crate::mast::BasicBlockNode`], all its operation- and decorator-related data is
+/// stored in the serialized [`MastForest`]'s `data` field at offset represented by the `offset`
+/// field. For all other variants of [`MastNode`], the `offset` field is guaranteed to be 0.
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs:15]: fixed

File: core/src/mast/serialization/info.rs [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+                Ok(MastNode::new_basic_block_with_decorators(operations, decorators))
+            }
+            MastNodeType::Join {
+                left_child_id,
+                right_child_id,
+            } => {
+                let left_child = MastNodeId::from_u32_safe(left_child_id, mast_forest)?;
+                let right_child = MastNodeId::from_u32_safe(right_child_id, mast_forest)?;
+
+                Ok(MastNode::new_join(left_child, right_child, mast_forest))
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs:58]: #1391

File: core/src/mast/serialization/info.rs [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+
+        if mast_node.digest() == self.digest {
+            Ok(mast_node)
+        } else {
+            Err(DeserializationError::InvalidValue(format!(
+                "MastNodeInfo's digest '{}' doesn't match deserialized MastNode's digest '{}'",
+                self.digest,
+                mast_node.digest()
+            )))
+        }
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs:96]: Also added a link to this comment in #1391

File: core/src/mast/serialization/info.rs [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+            MastNodeType::Loop { body_id: body } => Self::encode_u32_payload(*body),
+            MastNodeType::Block { offset, len } => Self::encode_u32_pair(*offset, *len),
+            MastNodeType::Call { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::SysCall { callee_id } => Self::encode_u32_payload(*callee_id),
+            MastNodeType::Dyn => [0; 8],
+            MastNodeType::External => [0; 8],
+        }
+    }
+
+    fn encode_u32_pair(left_child_id: u32, right_child_id: u32) -> [u8; 8] {
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs:256]: Oops I missed those; fixed

File: core/src/mast/serialization/basic_block_data_decoder.rs (hunk 12 above)
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_decoder.rs:34]: Done

File: core/src/mast/serialization/basic_block_data_builder.rs [hunk 18]
```diff
@@ -0,0 +1,187 @@
...
+}
+
+impl StringTableBuilder {
+    pub fn add_string(&mut self, string: &str) -> StringIndex {
+        if let Some(str_idx) = self.str_to_index.get(&Blake3_256::hash(string.as_bytes())) {
+            // return already interned string
+            *str_idx
+        } else {
+            // add new string to table
+            // NOTE: these string refs' offset will need to be shifted again in `into_buffer()`
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs:155]: yup, fixed

File: core/src/mast/serialization/basic_block_data_builder.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/basic_block_data_builder.rs]: Switched to `DataOffset`

File: core/src/mast/serialization/info.rs [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+            // significant bits. Also, the most significant byte of the right child is guaranteed to
+            // fit in 6 bits. Hence, we use big endian format for the right child id to simplify
+            // encoding and decoding.
+            let [msb, a, b, lsb] = right_child_id.to_be_bytes();
+
+            result[4] |= msb;
+            result[5] = a;
+            result[6] = b;
+            result[7] = lsb;
+        };
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Applied suggestion. This also panics if either ids doesn't fit in 30 bits (which will never be hit after #1392).

File: core/src/mast/serialization/info.rs [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+
+        result
+    }
+
+    fn encode_u32_payload(payload: u32) -> [u8; 8] {
+        let [payload_byte1, payload_byte2, payload_byte3, payload_byte4] = payload.to_be_bytes();
+
+        [0, payload_byte1, payload_byte2, payload_byte3, payload_byte4, 0, 0, 0]
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: > Another potential simplification is to put the payload in the lower 32 bits.

Applied suggestion

File: core/src/mast/serialization/info.rs [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+            // Tag is always placed in the first four bytes
+            let discriminant = self.discriminant();
+            assert!(discriminant <= 0b1111);
+            serialized_bytes[0] |= discriminant << 4;
+
+            serialized_bytes
+        };
+
+        serialized_bytes.write_into(target)
+    }
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Applied suggestion

File: core/src/mast/serialization/info.rs [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+}
+
+impl Deserializable for MastNodeType {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let bytes: [u8; 8] = source.read_array()?;
+
+        let tag = bytes[0] >> 4;
+
+        match tag {
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Applied suggestion

File: core/src/mast/serialization/info.rs [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+        let second = {
+            let mut second_be_bytes = [0_u8; 4];
+
+            second_be_bytes[0] = buffer[4] & 0b0011_1111;
+            second_be_bytes[1] = buffer[5];
+            second_be_bytes[2] = buffer[6];
+            second_be_bytes[3] = buffer[7];
+
+            u32::from_be_bytes(second_be_bytes)
+        };
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Applied suggestion, but as commented above, this now assumes that the input `payload` is 60 bits

File: core/src/mast/serialization/info.rs [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+
+        (first, second)
+    }
+
+    pub fn decode_u32_payload(payload: [u8; 8]) -> u32 {
+        let payload_be_bytes = [payload[1], payload[2], payload[3], payload[4]];
+
+        u32::from_be_bytes(payload_be_bytes)
+    }
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Applied suggestion, but also returns an error if the payload doesn't fit into a `u32`

**Comment by plafer** [comment]
Also added a few tests for `MastNodeType` serialization/deserialization.

File: core/src/mast/serialization/info.rs [hunk 25]
```diff
@@ -0,0 +1,407 @@
...
+                left_value
+            );
+        }
+
+        if right_value.leading_zeros() < 2 {
+            panic!(
+                "MastNodeType::encode_u32_pair: right value doesn't fit in 30 bits: {}",
+                left_value
+            );
+        }
```
[the remark this answers is not in the document]
Comment (plafer) [reply core/src/mast/serialization/info.rs]: Right... Fixed

File: core/src/mast/serialization/info.rs [hunk 26]
```diff
@@ -0,0 +1,404 @@
...
+    /// # Panics
+    /// - Panics if either `left_value` or `right_value` doesn't fit in 30 bits.
+    fn encode_u32_pair(left_value: u32, right_value: u32) -> u64 {
+        assert!(
+            left_value.leading_zeros() < 2,
+            "MastNodeType::encode_u32_pair: left value doesn't fit in 30 bits: {}",
+            left_value
+        );
+        assert!(
+            right_value.leading_zeros() < 2,
```
Comment (bobbinth) [inline core/src/mast/serialization/info.rs:254]: Should the conditions not be `left_value.leading_zeros() >= 2`?

**Review by bitwalker — APPROVED** [review]
Looks good!
